{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 7 Assignment\n",
    "\n",
    "_MkKinney 6.1_\n",
    "\n",
    "This week has been all about getting information off the internet both in structured data formats (CSV, JSON, etc) as well as HTML.  For these exercises, we're going to use two practical examples of fetching data from web pages to show how to use Pandas and BeautifulSoup to extract structured information from the web.\n",
    "\n",
    "---\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 33.1 Parsing a list in HTML\n",
    "\n",
    "Go to the Banner Health Price Transparency Page: https://www.bannerhealth.com/patients/billing/pricing-resources/hospital-price-transparency\n",
    "\n",
    "Notice that there is a list of hospitals and the city they are in.  We want to parse the underlying HTML to create a list of all the hospitals along with which city they're in.\n",
    "\n",
    "```json\n",
    "[\n",
    "    [\"Banner - University Medical Center Phoenix\", \"Arizona\"],\n",
    "    [\"Banner - University Medical Center South \", \"Arizona\"],\n",
    "    ...\n",
    "]\n",
    "```\n",
    "\n",
    "To examine the underlying HTML code, you can use Chrome, right-click, and choose **Inspect**.\n",
    "\n",
    "For reference, the documentation for BeautifulSoup is here: https://www.crummy.com/software/BeautifulSoup/bs4/doc/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the cell below, create a function called **parse_banner(url)** that takes as it's one parameter the URL of the webpage to be parsed for links.  Make sure you include docstrings and a good test case using hte URL provided above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "\n",
    "# Note that you'll need to fetch the data using the following syntax to include headers\n",
    "# that make the web server think you're a real web browser.\n",
    "headers = { \"User-Agent\": \"Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36\" }\n",
    "response = requests.get('https://www.bannerhealth.com/patients/billing/pricing-resources/hospital-price-transparency', headers=headers)\n",
    "s = BeautifulSoup(response.text, 'html.parser')\n",
    "div = s.find_all('div', {\"class\":\"col-md-8\"})[0]\n",
    "def parse_banner(url):\n",
    "    for hospital_list in div.find_all('ul'):\n",
    "        state = hospital_list.previous_sibling.previous_sibling.string\n",
    "        for hospital in hospital_list.find_all('li'):\n",
    "            print(state, hospital.text)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding tests in NoName\n"
     ]
    }
   ],
   "source": [
    "import doctest\n",
    "doctest.run_docstring_examples(parse_banner, globals(), verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Arizona Banner - University Medical Center Phoenix\n",
      "Arizona Banner - University Medical Center South \n",
      "Arizona Banner - University Medical Center Tucson\n",
      "Arizona Banner Baywood Medical Center \n",
      "Arizona Banner Behavioral Health Hospital\n",
      "Arizona Banner Boswell Medical Center\n",
      "Arizona Banner Casa Grande Medical Center\n",
      "Arizona Banner Del E. Webb Medical Center\n",
      "Arizona Banner Desert Medical Center/Cardon Children's Medical Center  \n",
      "Arizona Banner Estrella Medical Center\n",
      "Arizona Banner Gateway Medical Center/Banner MD Anderson Cancer Center\n",
      "Arizona Banner Goldfield Medical Center  \n",
      "Arizona Banner Heart Hospital\n",
      "Arizona Banner Ironwood Medical Center\n",
      "Arizona Banner Ocotillo Medical Center\n",
      "Arizona Banner Payson Medical Center\n",
      "Arizona Banner Rehabilitation Hospitals\n",
      "Arizona Banner Thunderbird Medical Center\n",
      "Arizona Page Hospital\n",
      "California Banner Lassen Medical Center\n",
      "Colorado Banner Fort Collins Medical Center\n",
      "Colorado McKee Medical Center\n",
      "Colorado North Colorado Medical Center\n",
      "Colorado Sterling Regional Medical Center\n",
      "Colorado East Morgan County Hospital\n",
      "Nebraska Ogallala Community Hospital\n",
      "Nevada Banner Churchill Community Hospital\n",
      "Wyoming Banner Wyoming Medical Center Central Campus\n",
      "Wyoming Banner Wyoming Medical Center East Campus\n",
      "Wyoming Community Hospital\n",
      "Wyoming Washakie Medical Center\n",
      "Wyoming Platte County Memorial Hospital\n",
      "State-Specific Regulations Arizona\n",
      "State-Specific Regulations California\n",
      "State-Specific Regulations Colorado\n",
      "State-Specific Regulations Nebraska\n",
      "State-Specific Regulations Nevada\n",
      "State-Specific Regulations Wyoming\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "object of type 'NoneType' has no len()",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-b50f7d3d19d6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mbanner\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse_banner\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'https://www.bannerhealth.com/patients/billing/pricing-resources/hospital-price-transparency'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbanner\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m38\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Length of result should have been 38, but {} returned.'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbanner\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;32massert\u001b[0m \u001b[0mbanner\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'Arizona'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Wrong data found in the first result item: {}'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbanner\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: object of type 'NoneType' has no len()"
     ]
    }
   ],
   "source": [
    "banner = parse_banner('https://www.bannerhealth.com/patients/billing/pricing-resources/hospital-price-transparency')\n",
    "assert len(banner)==38, 'Length of result should have been 38, but {} returned.'.format(len(banner))\n",
    "assert banner[0][1]=='Arizona', 'Wrong data found in the first result item: {}'.format(banner[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## 33.2 Using a REST API (from GitHub.com)\n",
    "\n",
    "Many websites provide something called a REST API to access information from their site programatically, rather than relying on HTML.  One example is GitHub.com, whose API allows you do to things like \"list all the public repositories for a user.\"\n",
    "\n",
    "The documentation for GitHub.com's REST API can be found here: https://docs.github.com/en/rest/guides/getting-started-with-the-rest-api\n",
    "\n",
    "Create a function called **repo_summary(user)** that takes a GitHub.com user name as it's parameter and retrieves a list of all the repositories you can see for that user.  The specific documentation for the this kind of request can be found here: https://docs.github.com/en/rest/reference/repos#list-repositories-for-a-user. Make sure your function is well documented with a docstring and includes a simple test to verify that you get back 12 repositories when querying for the repositories for user **paulboal**.\n",
    "\n",
    "I've provided a related example to help you out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This information is about paulboal. His website is www.amitechsolutions.com.\n"
     ]
    }
   ],
   "source": [
    "# Example -- this example of code shows how to get basic information on the user paulboal\n",
    "# For your solution, make sure you meet the requirements in the instructions above.\n",
    "\n",
    "import requests\n",
    "\n",
    "response = requests.get('https://api.github.com/users/paulboal')\n",
    "data = response.json()\n",
    "\n",
    "print('This information is about {}. His website is {}.'.format(data.get('login'), data.get('blog')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your code Here\n",
    "def repo_summary(user):\n",
    "    response = requests.get('https://api.github.com/users/'+user+'/repos')\n",
    "    data = response.json()\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding tests in NoName\n"
     ]
    }
   ],
   "source": [
    "import doctest\n",
    "doctest.run_docstring_examples(repo_summary, globals(), verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "repos = repo_summary('paulboal')\n",
    "assert len(repos)==12, 'Expecing 12, but {} were found'.format(len(repos))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "### 33.3 Find Something of Your Own\n",
    "\n",
    "Do some web searches and find an HTML page with some data that is interesting to something you're studying.  You can extract and parse that information using either BeautifulSoup or Pandas.  If you're using Pandas, then do something interesting to format and structure your data.  If you're using BeautifulSoup, you'll just need to do the work of parsing the data out of HTML -- that's hard enough!\n",
    "\n",
    "You don't need to build this as a function.  Just use notebook cells as I've done above.  You will be graded based on _style_.  Use variable names that make sense for your problem / solution. Cleanup anything you don't need before you submit your work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your Code Here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "tables = pd.read_html('https://www.arabidopsis.org/servlets/TairObject?id=28217&type=locus')\n",
    "len(tables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**************TABLE 0\n",
      "                  0                 1                        2   3\n",
      "0  Locus: AT1G19340  Locus: AT1G19340  What's new on this page NaN\n",
      "**************TABLE 1\n",
      "                                                    0  \\\n",
      "0                          Representative Gene  Model   \n",
      "1                                     Gene Model Type   \n",
      "2                                        Other names:   \n",
      "3                                         Description   \n",
      "4                                    Map Detail Image   \n",
      "5                                         Annotations   \n",
      "6                                            Sequence   \n",
      "7                                            RNA Data   \n",
      "8                                                 NaN   \n",
      "9                                  One-channel Arrays   \n",
      "10                                                NaN   \n",
      "11                             Associated Transcripts   \n",
      "12                                    BAR eFP Browser   \n",
      "13                                         Chromosome   \n",
      "14                                       Protein Data   \n",
      "15                               Search Gene Families   \n",
      "16                                  Phylogenetic Tree   \n",
      "17  PANTHER 11 Plant Homologs  Showing 7 of 7 entries   \n",
      "18                                      Map Locations   \n",
      "19                                          Map Links   \n",
      "20  Polymorphism Showing 15  of 228 entries (see all)   \n",
      "21      Germplasm Showing 15  of 22 entries (see all)   \n",
      "22                    Clones (showing 5 of 5 entries)   \n",
      "23                                      External Link   \n",
      "24  Community Comments (shows only the most recent...   \n",
      "25                                                NaN   \n",
      "26                                                NaN   \n",
      "27                                 Date last modified   \n",
      "28                                     TAIR Accession   \n",
      "\n",
      "                                                    1  \\\n",
      "0                                                 NaN   \n",
      "1                                                 NaN   \n",
      "2                                                 NaN   \n",
      "3                                                 NaN   \n",
      "4                                                 NaN   \n",
      "5                                                 NaN   \n",
      "6                                                 NaN   \n",
      "7                                                 NaN   \n",
      "8                                                 NaN   \n",
      "9                                                 NaN   \n",
      "10                                                NaN   \n",
      "11                                                NaN   \n",
      "12                                                NaN   \n",
      "13                                                NaN   \n",
      "14                                                NaN   \n",
      "15                                                NaN   \n",
      "16                                                NaN   \n",
      "17                                                NaN   \n",
      "18                                                NaN   \n",
      "19                                                NaN   \n",
      "20                                                NaN   \n",
      "21                                                NaN   \n",
      "22                                                NaN   \n",
      "23                                                NaN   \n",
      "24  Community Comments (shows only the most recent...   \n",
      "25                                                NaN   \n",
      "26                                                NaN   \n",
      "27                                                NaN   \n",
      "28                                                NaN   \n",
      "\n",
      "                                                    2  \n",
      "0                                         AT1G19340.1  \n",
      "1                                      protein_coding  \n",
      "2                                                 NaN  \n",
      "3   Methyltransferase MT-A70 family protein;(sourc...  \n",
      "4            Center on AT1G19340  |  Full-screen view  \n",
      "5   category  relationship  type keyword GO Biolog...  \n",
      "6   full length CDS  full length genomic  full len...  \n",
      "7                                                 NaN  \n",
      "8                                                 NaN  \n",
      "9   array elementname avg. signalintensity  (std. ...  \n",
      "10                                                NaN  \n",
      "11   type  number associated  EST  ( 8 )  cDNA  ( 3 )  \n",
      "12            Data Source  Powered by BAR Webservices  \n",
      "13                                                  1  \n",
      "14  name  length(aa)  molecular weight  isoelectri...  \n",
      "15  EnsemblPlants  PhyloGenes  InParanoid Ortholog...  \n",
      "16                                                NaN  \n",
      "17  Organism  UniProt Accession  Gene  Functional ...  \n",
      "18  chrom  map  map type coordinates  orientation ...  \n",
      "19         MapViewer Sequence Viewer GBrowse  JBrowse  \n",
      "20  name type polymorphism site  gene names  allel...  \n",
      "21  name  polymorphisms  background  stock name  S...  \n",
      "22  Clone Name  Aliases  GenBankAccession  VectorT...  \n",
      "23  Epigenomics  Danforth Center Small RNA/PARE/Me...  \n",
      "24  Community Comments (shows only the most recent...  \n",
      "25                                                NaN  \n",
      "26                                                NaN  \n",
      "27                                         2013-02-11  \n",
      "28                                      Locus:2016462  \n",
      "**************TABLE 2\n",
      "Empty DataFrame\n",
      "Columns: [AT1G19340.1]\n",
      "Index: []\n",
      "**************TABLE 3\n",
      "                          category         Unnamed: 1  \\\n",
      "0            GO Biological Process                NaN   \n",
      "1            GO Cellular Component                NaN   \n",
      "2            GO Cellular Component                NaN   \n",
      "3            GO Molecular Function                NaN   \n",
      "4  Growth and Developmental Stages                NaN   \n",
      "5                  Plant structure                NaN   \n",
      "6                Annotation Detail  Annotation Detail   \n",
      "\n",
      "           relationship  type         Unnamed: 3  \\\n",
      "0  acts upstream of or within                NaN   \n",
      "1                is active in                NaN   \n",
      "2                  located in                NaN   \n",
      "3                     enables                NaN   \n",
      "4            expressed during                NaN   \n",
      "5                expressed in                NaN   \n",
      "6           Annotation Detail  Annotation Detail   \n",
      "\n",
      "                                             keyword  \n",
      "0  embryo development, methylation, response to o...  \n",
      "1                                            nucleus  \n",
      "2                                            nucleus  \n",
      "3   methyltransferase activity, nucleic acid binding  \n",
      "4  flowering stage, petal differentiation and exp...  \n",
      "5  collective leaf structure, flower, flower pedi...  \n",
      "6                                  Annotation Detail  \n",
      "**************TABLE 4\n",
      "  array elementname avg. signalintensity  (std. error)  \\\n",
      "0         260669_AT                    44.308  (0.554)   \n",
      "\n",
      "  avg. signalpercentile  (std. error)  \n",
      "0                     37.750  (0.250)  \n",
      "**************TABLE 5\n",
      "   type number associated\n",
      "0   EST             ( 8 )\n",
      "1  cDNA             ( 3 )\n",
      "**************TABLE 6\n",
      "          name  Unnamed: 1  length(aa)  Unnamed: 3  molecular weight  \\\n",
      "0  AT1G19340.1         NaN       414.0         NaN           47892.8   \n",
      "1          NaN         NaN         NaN         NaN               NaN   \n",
      "2          NaN         NaN         NaN         NaN               NaN   \n",
      "\n",
      "   Unnamed: 5  isoelectric point  Unnamed: 7  \\\n",
      "0         NaN               6.51         NaN   \n",
      "1         NaN                NaN         NaN   \n",
      "2         NaN                NaN         NaN   \n",
      "\n",
      "                        INTERPRO domains  \n",
      "0  DNA_methylase_N6_adenine_CS:IPR002052  \n",
      "1                  MT-A70-like:IPR007757  \n",
      "2         SAM-dependent_MTases:IPR029063  \n",
      "**************TABLE 7\n",
      "                                 0   1           2\n",
      "0                    EnsemblPlants NaN  PhyloGenes\n",
      "1       InParanoid Ortholog Groups NaN   PhylomeDB\n",
      "2                          PANTHER NaN   Phytozome\n",
      "3  PGDD duplications and orthologs NaN       PLAZA\n",
      "**************TABLE 8\n",
      "    0   1\n",
      "0 NaN NaN\n",
      "**************TABLE 9\n",
      "                                       Organism UniProt Accession  \\\n",
      "0  Brachypodium distachyon (Purple false brome)            I1H8J3   \n",
      "1                         Glycine max (Soybean)            K7M5H2   \n",
      "2                         Glycine max (Soybean)            K7MNU4   \n",
      "3                  Physcomitrella patens (Moss)            A9RWK4   \n",
      "4        Populus trichocarpa (Black Cottonwood)            U5FSB1   \n",
      "5                 Solanum lycopersicum (Tomato)            K4BV74   \n",
      "6                        Vitis vinifera (Grape)            F6H041   \n",
      "\n",
      "                Gene Functional Information  \n",
      "0       BRADI1G71260      Search in QuickGO  \n",
      "1      GLYMA14G08331      Search in QuickGO  \n",
      "2      GLYMA17G36721      Search in QuickGO  \n",
      "3  PHYPADRAFT_206270      Search in QuickGO  \n",
      "4  POPTR_0014S04090G      Search in QuickGO  \n",
      "5   SOLYC04G079950.2      Search in QuickGO  \n",
      "6  VIT_18S0001G11990      Search in QuickGO  \n",
      "**************TABLE 10\n",
      "   chrom  map      map type             coordinates orientation  attrib  \\\n",
      "0      1  AGI  nuc_sequence  6684467  - 6687116  bp     forward     NaN   \n",
      "\n",
      "  Unnamed: 6  \n",
      "0    details  \n",
      "**************TABLE 11\n",
      "           name  Unnamed: 1       type  Unnamed: 3 polymorphism site  \\\n",
      "0   FLAG_123E08         NaN  insertion         NaN          promoter   \n",
      "1   GABI_405B07         NaN  insertion         NaN             3'UTR   \n",
      "2   GABI_456D02         NaN  insertion         NaN             3'UTR   \n",
      "3   GABI_551D01         NaN  insertion         NaN            intron   \n",
      "4   GABI_551F08         NaN  insertion         NaN            intron   \n",
      "5   GABI_552F07         NaN  insertion         NaN            intron   \n",
      "6   GABI_552F12         NaN  insertion         NaN            intron   \n",
      "7   GABI_554B10         NaN  insertion         NaN            intron   \n",
      "8   GABI_556A07         NaN  insertion         NaN            intron   \n",
      "9   GABI_556A09         NaN  insertion         NaN            intron   \n",
      "10  GABI_556C03         NaN  insertion         NaN            intron   \n",
      "11  GABI_556C06         NaN  insertion         NaN            intron   \n",
      "12  GABI_556D04         NaN  insertion         NaN            intron   \n",
      "13  GABI_556E05         NaN  insertion         NaN            intron   \n",
      "14  GABI_556E11         NaN  insertion         NaN            intron   \n",
      "\n",
      "    Unnamed: 5   gene names  Unnamed: 7 allele type  \n",
      "0          NaN  AT1G19340.1         NaN     unknown  \n",
      "1          NaN  AT1G19340.1         NaN     unknown  \n",
      "2          NaN  AT1G19340.1         NaN     unknown  \n",
      "3          NaN  AT1G19340.1         NaN     unknown  \n",
      "4          NaN  AT1G19340.1         NaN     unknown  \n",
      "5          NaN  AT1G19340.1         NaN     unknown  \n",
      "6          NaN  AT1G19340.1         NaN     unknown  \n",
      "7          NaN  AT1G19340.1         NaN     unknown  \n",
      "8          NaN  AT1G19340.1         NaN     unknown  \n",
      "9          NaN  AT1G19340.1         NaN     unknown  \n",
      "10         NaN  AT1G19340.1         NaN     unknown  \n",
      "11         NaN  AT1G19340.1         NaN     unknown  \n",
      "12         NaN  AT1G19340.1         NaN     unknown  \n",
      "13         NaN  AT1G19340.1         NaN     unknown  \n",
      "14         NaN  AT1G19340.1         NaN     unknown  \n",
      "**************TABLE 12\n",
      "                name        Unnamed: 1  \\\n",
      "0        SALK_070229               NaN   \n",
      "1             images            images   \n",
      "2     None available    None available   \n",
      "3         phenotypes        phenotypes   \n",
      "4     None available    None available   \n",
      "..               ...               ...   \n",
      "85            images            images   \n",
      "86    None available    None available   \n",
      "87        phenotypes        phenotypes   \n",
      "88    None available    None available   \n",
      "89  (European Users)  (European Users)   \n",
      "\n",
      "                                        polymorphisms        Unnamed: 3  \\\n",
      "0   SALK_070229.55.50.x; SALKseq_070229.0; SALKseq...               NaN   \n",
      "1                                              images            images   \n",
      "2                                      None available    None available   \n",
      "3                                          phenotypes        phenotypes   \n",
      "4                                      None available    None available   \n",
      "..                                                ...               ...   \n",
      "85                                             images            images   \n",
      "86                                     None available    None available   \n",
      "87                                         phenotypes        phenotypes   \n",
      "88                                     None available    None available   \n",
      "89                                   (European Users)  (European Users)   \n",
      "\n",
      "          background        Unnamed: 5        stock name  \n",
      "0                NaN               NaN       SALK_070229  \n",
      "1             images            images            images  \n",
      "2     None available    None available    None available  \n",
      "3         phenotypes        phenotypes        phenotypes  \n",
      "4     None available    None available    None available  \n",
      "..               ...               ...               ...  \n",
      "85            images            images            images  \n",
      "86    None available    None available    None available  \n",
      "87        phenotypes        phenotypes        phenotypes  \n",
      "88    None available    None available    None available  \n",
      "89  (European Users)  (European Users)  (European Users)  \n",
      "\n",
      "[90 rows x 7 columns]\n",
      "**************TABLE 13\n",
      "      Clone Name   Aliases GenBankAccession VectorType Stock Number\n",
      "0   CATMA1A18370       NaN              NaN    plasmid     CD253645\n",
      "1  RAFL17-42-J02       NaN         AK117739    plasmid          NaN\n",
      "2       21403662  122669.0         AY084952        NaN          NaN\n",
      "3       BT012063       NaN         BT012063        NaN          NaN\n",
      "4  CATMA1A18370D       NaN              NaN    plasmid          NaN\n",
      "5            NaN       NaN              NaN        NaN          NaN\n",
      "**************TABLE 14\n",
      "                                                    0\n",
      "0                                         Epigenomics\n",
      "1          Danforth Center Small RNA/PARE/Methylation\n",
      "2                             Expression/Localization\n",
      "3                                              AraQTL\n",
      "4                                            ATTED-II\n",
      "5                                         eFP Browser\n",
      "6                  Eukaryotic Promoter Database (EPD)\n",
      "7                         Genevisible Expression Data\n",
      "8   The Subcellular Location of Proteins in Arabid...\n",
      "9              TraVA:Transcription Variation Analysis\n",
      "10  Gene Families  Links moved to Gene Families ba...\n",
      "11                                           Genomics\n",
      "12                                            AceView\n",
      "13                                            Gramene\n",
      "14                                               MIPS\n",
      "15                                   NCBI-Entrez Gene\n",
      "16                                   Thalemine at BAR\n",
      "17                                       Interactions\n",
      "18                                            BioGRID\n",
      "19                                              Other\n",
      "20                                         BAR ePlant\n",
      "21                           BAR ePlant Molecule View\n",
      "22                     Polymorphisms/Alleles/Variants\n",
      "23                                    Salk SNP Viewer\n",
      "24                                      T-DNA Express\n",
      "25                                         Proteomics\n",
      "26               AlphaFold Protein Structure Database\n",
      "27                                         AtProteome\n",
      "28  Functional Analysis Tools for Post-Translation...\n",
      "29                            Plant Proteome Database\n",
      "30                                   Plant PTM Viewer\n",
      "31                                           Reagents\n",
      "32                                Agrisera (antibody)\n",
      "33                                 PhytoAB (antibody)\n",
      "34       RIKEN BioResource Research Center (Seed/DNA)\n"
     ]
    }
   ],
   "source": [
    "for index,table in enumerate(tables):\n",
    "    print(\"**************TABLE {}\".format(index))\n",
    "    print(table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                    0  \\\n",
      "0                          Representative Gene  Model   \n",
      "1                                     Gene Model Type   \n",
      "2                                        Other names:   \n",
      "3                                         Description   \n",
      "4                                    Map Detail Image   \n",
      "5                                         Annotations   \n",
      "6                                            Sequence   \n",
      "7                                            RNA Data   \n",
      "8                                                 NaN   \n",
      "9                                  One-channel Arrays   \n",
      "10                                                NaN   \n",
      "11                             Associated Transcripts   \n",
      "12                                    BAR eFP Browser   \n",
      "13                                         Chromosome   \n",
      "14                                       Protein Data   \n",
      "15                               Search Gene Families   \n",
      "16                                  Phylogenetic Tree   \n",
      "17  PANTHER 11 Plant Homologs  Showing 7 of 7 entries   \n",
      "18                                      Map Locations   \n",
      "19                                          Map Links   \n",
      "20  Polymorphism Showing 15  of 228 entries (see all)   \n",
      "21      Germplasm Showing 15  of 22 entries (see all)   \n",
      "22                    Clones (showing 5 of 5 entries)   \n",
      "23                                      External Link   \n",
      "24  Community Comments (shows only the most recent...   \n",
      "25                                                NaN   \n",
      "26                                                NaN   \n",
      "27                                 Date last modified   \n",
      "28                                     TAIR Accession   \n",
      "\n",
      "                                                    1  \\\n",
      "0                                                 NaN   \n",
      "1                                                 NaN   \n",
      "2                                                 NaN   \n",
      "3                                                 NaN   \n",
      "4                                                 NaN   \n",
      "5                                                 NaN   \n",
      "6                                                 NaN   \n",
      "7                                                 NaN   \n",
      "8                                                 NaN   \n",
      "9                                                 NaN   \n",
      "10                                                NaN   \n",
      "11                                                NaN   \n",
      "12                                                NaN   \n",
      "13                                                NaN   \n",
      "14                                                NaN   \n",
      "15                                                NaN   \n",
      "16                                                NaN   \n",
      "17                                                NaN   \n",
      "18                                                NaN   \n",
      "19                                                NaN   \n",
      "20                                                NaN   \n",
      "21                                                NaN   \n",
      "22                                                NaN   \n",
      "23                                                NaN   \n",
      "24  Community Comments (shows only the most recent...   \n",
      "25                                                NaN   \n",
      "26                                                NaN   \n",
      "27                                                NaN   \n",
      "28                                                NaN   \n",
      "\n",
      "                                                    2  \n",
      "0                                         AT1G19340.1  \n",
      "1                                      protein_coding  \n",
      "2                                                 NaN  \n",
      "3   Methyltransferase MT-A70 family protein;(sourc...  \n",
      "4            Center on AT1G19340  |  Full-screen view  \n",
      "5   category  relationship  type keyword GO Biolog...  \n",
      "6   full length CDS  full length genomic  full len...  \n",
      "7                                                 NaN  \n",
      "8                                                 NaN  \n",
      "9   array elementname avg. signalintensity  (std. ...  \n",
      "10                                                NaN  \n",
      "11   type  number associated  EST  ( 8 )  cDNA  ( 3 )  \n",
      "12            Data Source  Powered by BAR Webservices  \n",
      "13                                                  1  \n",
      "14  name  length(aa)  molecular weight  isoelectri...  \n",
      "15  EnsemblPlants  PhyloGenes  InParanoid Ortholog...  \n",
      "16                                                NaN  \n",
      "17  Organism  UniProt Accession  Gene  Functional ...  \n",
      "18  chrom  map  map type coordinates  orientation ...  \n",
      "19         MapViewer Sequence Viewer GBrowse  JBrowse  \n",
      "20  name type polymorphism site  gene names  allel...  \n",
      "21  name  polymorphisms  background  stock name  S...  \n",
      "22  Clone Name  Aliases  GenBankAccession  VectorT...  \n",
      "23  Epigenomics  Danforth Center Small RNA/PARE/Me...  \n",
      "24  Community Comments (shows only the most recent...  \n",
      "25                                                NaN  \n",
      "26                                                NaN  \n",
      "27                                         2013-02-11  \n",
      "28                                      Locus:2016462  \n"
     ]
    }
   ],
   "source": [
    "print(tables[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Check your work above\n",
    "\n",
    "If you didn't get them all correct, take a few minutes to think through those that aren't correct.\n",
    "\n",
    "\n",
    "## Submitting Your Work\n",
    "\n",
    "In order to submit your work, you'll need to use the `git` command line program to **add** your homework file (this file) to your local repository, **commit** your changes to your local repository, and then **push** those changes up to github.com.  From there, I'll be able to **pull** the changes down and do my grading.  I'll provide some feedback, **commit** and **push** my comments back to you.  Next week, I'll show you how to **pull** down my comments.\n",
    "\n",
    "To run through everything one last time and submit your work:\n",
    "1. Use the `Kernel` -> `Restart Kernel and Run All Cells` menu option to run everything from top to bottom and stop here.\n",
    "2. Follow the instruction on the prompt below to either ssave and submit your work, or continue working.\n",
    "\n",
    "If anything fails along the way with this submission part of the process, let me know.  I'll help you troubleshoort."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "\n",
      "Are you ready to submit your work?\n",
      "1. Click the Save icon (or do Ctrl-S / Cmd-S)\n",
      "2. Type \"yes\" or \"no\" below\n",
      "3. Press Enter\n",
      "\n",
      " yes\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[main d88bc5b] Submitting the week 7 programming exercises\n",
      " 2 files changed, 297 insertions(+), 8 deletions(-)\n",
      " create mode 100644 week07/week07_assignment_2.ipynb\n",
      "Counting objects: 5, done.\n",
      "Delta compression using up to 2 threads.\n",
      "Compressing objects: 100% (5/5), done.\n",
      "Writing objects: 100% (5/5), 3.66 KiB | 3.66 MiB/s, done.\n",
      "Total 5 (delta 2), reused 0 (delta 0)\n",
      "remote: Resolving deltas: 100% (2/2), completed with 2 local objects.\u001b[K\n",
      "To github.com:paulboal/hds5210-2022.git\n",
      "   fc2200c..d88bc5b  main -> main\n"
     ]
    }
   ],
   "source": [
    "a=input('''\n",
    "Are you ready to submit your work?\n",
    "1. Click the Save icon (or do Ctrl-S / Cmd-S)\n",
    "2. Type \"yes\" or \"no\" below\n",
    "3. Press Enter\n",
    "\n",
    "''')\n",
    "\n",
    "if a=='yes':\n",
    "    !git add week07_assignment_2.ipynb\n",
    "    !git commit -a -m \"Submitting the week 7 programming exercises\"\n",
    "    !git push\n",
    "else:\n",
    "    print('''\n",
    "    \n",
    "OK. We can wait.\n",
    "''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
